### PR TITLE
Incorporate arg-parsing into the library

### PR DESCRIPTION
**Pull request by jszaday**
Proof of concept for incorporating arg-parsing into the library, remaining tasks:

- [x]  Track origin of parameters (e.g., file or command-line).
- [x]  Add support for boolean arguments (store true/false).
- [x] Ensure origin is preserved beyond `main`.
- [ ] Update documentation to reflect changes introduced by this PR.

### CODE REVIEW

File: .github/workflows/smp.yml [hunk 1]
```diff
@@ -49,11 +49,11 @@ jobs:
         make
         ./charmrun +p1 ./Gravity -f ../inputgen/1k.tipsy ++ppn 1 +setcpuaffinity ++local -u 2
         ./charmrun +p1 ./Gravity -f ../inputgen/1k.tipsy ++ppn 1 +setcpuaffinity ++local -u 2 -n 1 -p 1 -d sfc -t kd
-        ./charmrun +p4 ./Gravity -f ../inputgen/1k.tipsy ++ppn 4 +setcpuaffinity ++local -u 2 -t bin
+        ./charmrun +p4 ./Gravity -f ../inputgen/1k.tipsy ++ppn 4 +setcpuaffinity ++local -u 2 -t binoct
```
Comment (jszaday) [inline .github/workflows/smp.yml:52]: Where the previous arg-parser would allow invalid values, the current one will fail here.

Comment (trquinn) [reply .github/workflows/smp.yml:52]: That is an acceptable outcome.

**Comment by trquinn** [comment]
I'm having trouble with Loadable::origin_of(): it's not finding a parameter that I specified in the "param" file.  Could you check that it does actually work?

**Comment by trquinn** [comment]
Loadable::origin_of() appears to be working.

**Review by trquinn — APPROVED** [review]

**Pull request merged**